kvserver: don't allow raft forwarding of lease requests

andreimatei opened the pull request:
This patch aims to improve the behavior in scenarios where a follower
replica is behind, unaware of the latest lease, and it tries to acquire
a lease in its ignorant state. That lease acquisition request is bound
to fail (because the lease that it's based on is stale), but while it
fails (or, rather, until the behind replica finds out that it failed)
local requests are blocked. This blocking can last for a very long time
in situations where a snapshot is needed to catch up the follower, and
the snapshot is queued up behind many other snapshots (e.g. after a node
has been down for a while and gets restarted).

This patch tries an opinionated solution: never allow followers to
acquire leases. If there is a leader, it's a better idea for the leader
to acquire the lease. The leader might have a lease anyway or, even if
it doesn't, having the leader acquire it saves a leadership transfer
(leadership follows the lease).
We change the proposal path to recognize lease requests and reject them
early if the current replica is a follower and the leader is known. The
rejection points to the leader, which causes the request that triggered
the lease acquisition to make its way to the leader and attempt to
acquire a lease over there.

The next commit takes this further by short-circuiting the lease
proposal even sooner - but that patch is more best-effort.

Fixes #37906

Release note (bug fix): A bug causing queries sent to a
freshly-restarted node to sometimes hang for a long time while the node
catches up with replication has been fixed.

andreimatei committed:
kvserver: plumb ctx to proposal buffer

Release note: None

andreimatei committed:
kvserver: don't allow raft forwarding of lease requests

This patch aims to improve the behavior in scenarios where a follower
replica is behind, unaware of the latest lease, and it tries to acquire
a lease in its ignorant state. That lease acquisition request is bound
to fail (because the lease that it's based on is stale), but while it
fails (or, rather, until the behind replica finds out that it failed)
local requests are blocked. This blocking can last for a very long time
in situations where a snapshot is needed to catch up the follower, and
the snapshot is queued up behind many other snapshots (e.g. after a node
has been down for a while and gets restarted).

This patch tries an opinionated solution: never allow followers to
acquire leases. If there is a leader, it's a better idea for the leader
to acquire the lease. The leader might have a lease anyway or, even if
it doesn't, having the leader acquire it saves a leadership transfer
(leadership follows the lease).
We change the proposal path to recognize lease requests and reject them
early if the current replica is a follower and the leader is known. The
rejection points to the leader, which causes the request that triggered
the lease acquisition to make its way to the leader and attempt to
acquire a lease over there.

Fixes #37906

As described in #37906, the badness caused by requests blocking behind a
doomed lease acq request could be reproduced with a 100-warehouse tpcc
workload (--no-wait) on a 3 node cluster. Turning off a node for 10
minutes and then turning it back on would result in the cluster being
wedged for a few minutes until all the snapshots are transferred. I've
verified that this patch fixes it.

Release note (bug fix): A bug causing queries sent to a
freshly-restarted node to sometimes hang for a long time while the node
catches up with replication has been fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55148)
<!-- Reviewable:end -->

andreimatei commented:
Patch is not quite ready and I will work on tests too, but I wanna see if it raises any discussion.
Particularly from @bdarnell; the approach in the first commit was discussed with the rest of the gang.

nvb reviewed (commented):
> Reviewed 5 of 5 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @knz, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfRweL-Qj6_XixvHXG:-MIfRweL-Qj6_XixvHXH:b1sdci3) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L163)):*
> ```Go
> 	// rejectProposalWithRedirectLocked rejects a proposal and redirects the
> 	// proposer to try it on another node. This is used to sometimes reject lease
> 	// acquisitions when the another replica is the leader; the intended
> ```

"the another"
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 411 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfSN1_2-3re0YlBQPM:-MIfSN1_2-3re0YlBQPN:b-yvjbtz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L411)):*
> ```Go
> 	ents := make([]raftpb.Entry, 0, used)
> 
> 	var follower bool
> ```

Give this block a comment.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 703 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfTCv54mguzJK2iZbn:-MIfTCv54mguzJK2iZbo:blmp4l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L703)):*
> ```Go
> }
> 
> // rejectProposalWithRedirectLocked is part of the proposer interface.
> ```

I think we also need to call `cleanupFailedProposalLocked` in this function before calling `finishApplication`. That will ensure that the proposal is removed from the `proposals` map and that it will release quota if it acquired any.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 708 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfTkr3AEG5Orr8IlmS:-MIfTkr3AEG5Orr8IlmT:b-xpodyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L708)):*
> ```Go
> ) {
> 	r := (*Replica)(rp)
> 	rangeDesc := r.Desc()
> ```

Doesn't this grab a read lock? Read locks aren't re-entrant. I think you want `descRLocked`
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 709 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfTzUD-6r7M7hJ7wru:-MIfTzUD-6r7M7hJ7wrv:bdvvjvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L709)):*
> ```Go
> 	r := (*Replica)(rp)
> 	rangeDesc := r.Desc()
> 	replicaDesc, err := r.getReplicaDescriptorRLocked()
> ```

If you just want the `StoreID`, why go through all this effort. Is `r.store.StoreID()` not sufficient?
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfTU7kAcuIXjgA8XLl:-MIfTU7kAcuIXjgA8XLm:b4mgpo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf_test.go#L74)):*
> ```Go
> }
> 
> func (t *testProposer) rejectProposalWithRedirectLocked(
> ```

It seems like we should be able to write some tests for this fairly easily.
___
*[pkg/kv/kvserver/replica_range_lease.go, line 212 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfVMHu-5K6eDmQ36v3:-MIfVMHu-5K6eDmQ36v4:b-b6pwrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e27baf663248d42d548cbf3a6d50a2ab6e0bc80b/pkg/kv/kvserver/replica_range_lease.go#L212)):*
> ```Go
> 	// every request from making it too far down the proposal path - in particular
> 	// to avoid the heartbeating of the liveness record done below.
> 	leader := p.repl.mu.leaderID
> ```

I'm not sold on this. We want to backport this PR, so we'd like it to be as small as possible. And we're planning on using `master` as a staging ground for the backport, so we shouldn't merge anything on top that we don't intend to backport in case it hides issues with the portion we do intend to backport.

So I'm going to assume that if this is in this PR, we want to backport it too. If so, then there's a high bar for changes, especially in such subtle areas as this. This is only an optimization, right? Have we seen that it's anything more than theoretical? Have we seen the liveness heartbeats discussed above cause issues? As of 20.2, liveness heartbeats are coalesced, so I'd be surprised if they were.

I'm primarily worried about the additional complexity this adds in exchange for the unclear optimization. The PR itself is already making a potentially destabilizing change, so I'm wary of making any issues that fall out from it harder to debug or increases the chance that we got it wrong.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
r2 looks good to me; i share Nathan's reservations about r3. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-:-MJ8z-br56RlDAOW3T-m:b6ee7x9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @knz, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 421 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MJ8yHJb938MPUqs2LKa:-MJ8yHJb938MPUqs2LKb:bdbvyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L421)):*
> ```Go
> 			leader = roachpb.ReplicaID(status.Lead)
> 			if leader == b.p.replicaID() {
> 				// !!! this assertion fails, apparently in single-replica groups, so I
> ```

Yeah, I remember that single-replica groups short-circuit a lot of logic and don't maintain a lot of things that look like invariants in multi-replica groups. I don't remember anything well enough to give you useful advice right now, though. 
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 463 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MJ8yA311HJ_bxdtkRjC:-MJ8yA311HJ_bxdtkRjD:b-clfv0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L463)):*
> ```Go
> 		// - if the leader is known, we reject this proposal and make sure the
> 		// request that needed the lease is redirected to the leaseholder;
> 		// - if the leader is not known, we don't do anything special here to
> ```

We don't *have* to do anything special here because it will get dropped later, but could we just reject the proposal here whether we know the leader or not? Would avoid questions like the one in the assertion above. 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @knz, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 463 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MJ8yA311HJ_bxdtkRjC:-MJ9IRxN4gelgIqpuVx8:b-l259ov) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L463)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We don't *have* to do anything special here because it will get dropped later, but could we just reject the proposal here whether we know the leader or not? Would avoid questions like the one in the assertion above. 
</blockquote></details>

I was going to respond that we don't want to immediately reject the proposal because the case where `!leaderKnown` is the case where we'd rather wait until the current election completes so we know who to redirect to, but now I'm confused. When will we ever be a follower and not know the current leader? I thought you'll always be a candidate/pre-candidate if you don't know the leader.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @knz, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 416 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MJCor_L1f-zBN8KuxR4:-MJCor_L1f-zBN8KuxR5:b3ugnl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L416)):*
> ```Go
> 	if raftGroup != nil {
> 		status := raftGroup.BasicStatus()
> 		follower = status.RaftState == raft.StateFollower
> ```

Should this be `state == follower` or `state != leader`? I think candidates and pre-candidates (and learners, although they just use StateFollower) should be handled the same way as followers here. 
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 463 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MJ8yA311HJ_bxdtkRjC:-MJCngZ30F1hFIEQCKec:b8jvqks) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L463)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I was going to respond that we don't want to immediately reject the proposal because the case where `!leaderKnown` is the case where we'd rather wait until the current election completes so we know who to redirect to, but now I'm confused. When will we ever be a follower and not know the current leader? I thought you'll always be a candidate/pre-candidate if you don't know the leader.
</blockquote></details>

The initial state of a raft instance is to be a follower with unknown leader. And there are a few ways to get back to that state. For example, if an election fails without a majority winner, the candidates become followers (with no leaders) as they back off before starting a new election. 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @knz, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 463 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MJ8yA311HJ_bxdtkRjC:-MJDSEpC7JgnPhw6UTwL:bab7mzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L463)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The initial state of a raft instance is to be a follower with unknown leader. And there are a few ways to get back to that state. For example, if an election fails without a majority winner, the candidates become followers (with no leaders) as they back off before starting a new election. 
</blockquote></details>

Thanks for the explanation, that's very helpful. Then I agree with your other comment that we should invert the `follower` boolean to be a `leader` boolean and then check `if !leader && p.Request.IsLeaseRequest() {`.

I still think we should only redirect if the leader is known, though, for the reasons listed in https://github.com/cockroachdb/cockroach/issues/37906#issuecomment-666493887.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @knz, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 463 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MJ8yA311HJ_bxdtkRjC:-MJYETLm93N-YB7BW6_X:b4ye1iv) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L463)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Thanks for the explanation, that's very helpful. Then I agree with your other comment that we should invert the `follower` boolean to be a `leader` boolean and then check `if !leader && p.Request.IsLeaseRequest() {`.

I still think we should only redirect if the leader is known, though, for the reasons listed in https://github.com/cockroachdb/cockroach/issues/37906#issuecomment-666493887.
</blockquote></details>

I don't feel strongly about what happens when there is no leader - it's a very rare case and either bouncing among random replicas or waiting synchronously would be acceptable to me. If we could get rid of the synchronous wait code altogether that would be nice, although I guess that's still useful in the normal cold start path. 


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
PTAL at the test

I hit a problem though: replicas on draining nodes will not take leases. They will become leaders, though, which can lead to deadlock since, if a leader is draining, nobody else will attempt to take the lease and it also won’t.
The logic is here: https://github.com/cockroachdb/cockroach/blob/785aea768a8c8bdf95f8743a74a14c973baf0f6e/pkg/kv/kvserver/replica_range_lease.go#L619-L623
I’m also thinking if this is already a deadlock condition: is there anything preventing all the nodes with replicas for a range from draining at the same time? If that happens, nobody will want the lease?
I'm thinking of allowing a leader to acquire the lease even if it's draining.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-:-MJdY2ft3sVbi23HIQMv:bnsbvdl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @knz, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfRweL-Qj6_XixvHXG:-MJYxDGX9LuE9pes7BsP:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L163)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"the another"
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 411 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfSN1_2-3re0YlBQPM:-MJYyF2WBEaidFaRt1GY:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L411)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Give this block a comment.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 416 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MJCor_L1f-zBN8KuxR4:-MJYxgE_5_Z1ewR4KbBt:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L416)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Should this be `state == follower` or `state != leader`? I think candidates and pre-candidates (and learners, although they just use StateFollower) should be handled the same way as followers here. 
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 463 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MJ8yA311HJ_bxdtkRjC:-MJYyK-S9m8tZAlY8DqZ:baxaqhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L463)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't feel strongly about what happens when there is no leader - it's a very rare case and either bouncing among random replicas or waiting synchronously would be acceptable to me. If we could get rid of the synchronous wait code altogether that would be nice, although I guess that's still useful in the normal cold start path. 
</blockquote></details>

I think I'm in the "let's wait" camp too. Seems cleaner to me... I don't really like `NotLeaseholderErrors` that don't have a redirection in them.
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 703 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfTCv54mguzJK2iZbn:-MJblVrE6YFOSALGfxxj:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L703)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think we also need to call `cleanupFailedProposalLocked` in this function before calling `finishApplication`. That will ensure that the proposal is removed from the `proposals` map and that it will release quota if it acquired any.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 708 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfTkr3AEG5Orr8IlmS:-MJbnEs8A76JMlAqnKtp:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L708)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Doesn't this grab a read lock? Read locks aren't re-entrant. I think you want `descRLocked`
</blockquote></details>

removed
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 709 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfTzUD-6r7M7hJ7wru:-MJbnFeL8p7O7MDRqJWw:bbvwp7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L709)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If you just want the `StoreID`, why go through all this effort. Is `r.store.StoreID()` not sufficient?
</blockquote></details>

done, 10x


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
> Reviewed 2 of 5 files at r1, 1 of 4 files at r4, 5 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @knz, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/client_raft_test.go, line 1157 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MK0j6Zy-tezjtez_tbZ:-MK0j6ZzDyUGvedAUbhM:b-ujmrte) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e727d5f42f1e732a84c285959c684acb2bfab/pkg/kv/kvserver/client_raft_test.go#L1157)):*
> ```Go
> 	log.Infof(ctx, "test: truncating log... done")
> 
> 	// !!! assert lease expired
> ```

Don't forget this. 


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-:-MKBo5520ZO5jh4qwXXv:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @knz, and @tbg)

___
*[pkg/kv/kvserver/client_raft_test.go, line 1054 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MKBm8vhBMrWo9DcF945:-MKBm8vhBMrWo9DcF946:b2nglp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e727d5f42f1e732a84c285959c684acb2bfab/pkg/kv/kvserver/client_raft_test.go#L1054)):*
> ```Go
> // the leader does not attempt to acquire a lease and, thus, does not block until
> // it figures out that it cannot, in fact, take the lease.
> // !!! more words
> ```

!!!
___
*[pkg/kv/kvserver/client_raft_test.go, line 1110 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MKBmpTOCHURiUQcdIqg:-MKBmpTOCHURiUQcdIqh:b-z2p6nt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e727d5f42f1e732a84c285959c684acb2bfab/pkg/kv/kvserver/client_raft_test.go#L1110)):*
> ```Go
> 	// Wait until the leadership is transferred away from the partitioned replica.
> 	testutils.SucceedsSoon(t, func() error {
> 		mtc.advanceClock(ctx)
> ```

How long does this test take to run? Does `advanceClock` actually speed up the election process? I wouldn't expect it to make a replica tick faster.
___
*[pkg/kv/kvserver/client_raft_test.go, line 1175 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MKBnHLO5G3xmDCg9uOQ:-MKBnHLO5G3xmDCg9uOR:b-x1sjdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e727d5f42f1e732a84c285959c684acb2bfab/pkg/kv/kvserver/client_raft_test.go#L1175)):*
> ```Go
> 
> 	// Now we're going to send a request to the behind replica, and we expect it
> 	// to not block; we expect a redirection to the leader.
> ```

Did you confirm that this blocked before the introduction of your fix?
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 418 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MKBnYAlAVd3Cg3cPmOx:-MKBnYAlAVd3Cg3cPmOy:b5fysa2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e727d5f42f1e732a84c285959c684acb2bfab/pkg/kv/kvserver/replica_proposal_buf.go#L418)):*
> ```Go
> 	// requests.
> 	var iAmTheLeader bool
> 	var leaderKnown bool
> ```

nit: `var leaderKnown, iAmTheLeader bool`


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
Should be ready to go. PTAL at the test.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-:-MKv1c1K5WZvVJ_f8qnN:b-zf8a1o)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bdarnell, @knz, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/client_raft_test.go, line 1054 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MKBm8vhBMrWo9DcF945:-MKv0fSz6-V2CZGvbCBv:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e727d5f42f1e732a84c285959c684acb2bfab/pkg/kv/kvserver/client_raft_test.go#L1054)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

!!!
</blockquote></details>

done
___
*[pkg/kv/kvserver/client_raft_test.go, line 1110 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MKBmpTOCHURiUQcdIqg:-MKv0gbf4p6sei7QU6GI:b-mxte9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e727d5f42f1e732a84c285959c684acb2bfab/pkg/kv/kvserver/client_raft_test.go#L1110)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

How long does this test take to run? Does `advanceClock` actually speed up the election process? I wouldn't expect it to make a replica tick faster.
</blockquote></details>

moved to TestServer. And tweaked some knobs. It takes ~3.5s
___
*[pkg/kv/kvserver/client_raft_test.go, line 1157 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MK0j6Zy-tezjtez_tbZ:-MKv0sA0DWpk1XUohcVA:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e727d5f42f1e732a84c285959c684acb2bfab/pkg/kv/kvserver/client_raft_test.go#L1157)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Don't forget this. 
</blockquote></details>

done
___
*[pkg/kv/kvserver/client_raft_test.go, line 1175 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MKBnHLO5G3xmDCg9uOQ:-MKv0tHj9lS72IZjlAYU:b2nih0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e727d5f42f1e732a84c285959c684acb2bfab/pkg/kv/kvserver/client_raft_test.go#L1175)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did you confirm that this blocked before the introduction of your fix?
</blockquote></details>

yup
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 421 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MJ8yHJb938MPUqs2LKa:-MKv14t1B2rKk3wHAq0D:bb7gzl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf.go#L421)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, I remember that single-replica groups short-circuit a lot of logic and don't maintain a lot of things that look like invariants in multi-replica groups. I don't remember anything well enough to give you useful advice right now, though. 
</blockquote></details>

gone
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 418 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MKBnYAlAVd3Cg3cPmOx:-MKv1JSd-LXTOxRss2Nl:b-7rl0jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b31e727d5f42f1e732a84c285959c684acb2bfab/pkg/kv/kvserver/replica_proposal_buf.go#L418)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `var leaderKnown, iAmTheLeader bool`
</blockquote></details>

done 
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfTU7kAcuIXjgA8XLl:-MKv1Py04V9-77wRn_JT:bj67hiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf_test.go#L74)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It seems like we should be able to write some tests for this fairly easily.
</blockquote></details>

added test
___
*[pkg/kv/kvserver/replica_range_lease.go, line 212 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfVMHu-5K6eDmQ36v3:-MKv1RxQ-Hr-dNk1XNJY:bh89j5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e27baf663248d42d548cbf3a6d50a2ab6e0bc80b/pkg/kv/kvserver/replica_range_lease.go#L212)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm not sold on this. We want to backport this PR, so we'd like it to be as small as possible. And we're planning on using `master` as a staging ground for the backport, so we shouldn't merge anything on top that we don't intend to backport in case it hides issues with the portion we do intend to backport.

So I'm going to assume that if this is in this PR, we want to backport it too. If so, then there's a high bar for changes, especially in such subtle areas as this. This is only an optimization, right? Have we seen that it's anything more than theoretical? Have we seen the liveness heartbeats discussed above cause issues? As of 20.2, liveness heartbeats are coalesced, so I'd be surprised if they were.

I'm primarily worried about the additional complexity this adds in exchange for the unclear optimization. The PR itself is already making a potentially destabilizing change, so I'm wary of making any issues that fall out from it harder to debug or increases the chance that we got it wrong.
</blockquote></details>

removed commit


<!-- Sent from Reviewable.io -->

tbg commented:
Test failure:             pkg/kv/kvserver/client_raft_test.go:1090:11: this value of err is never used (SA4006)

@nvanbenschoten you've got this PR, right?

nvb reviewed (commented):
> Reviewed 2 of 8 files at r6, 9 of 11 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @bdarnell, @knz, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/client_raft_test.go, line 1283 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLDt3lCAJD34XBhzyJp:-MLDt3lCAJD34XBhzyJq:buya36n) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d60b64af89c45cdebec183d2001003b28603051/pkg/kv/kvserver/client_raft_test.go#L1283)):*
> ```Go
> // the lease and then blocking traffic for a long time until they find out
> // whether they successfully took the lease or not.
> func TestRequestsOnLaggingReplicaXXX(t *testing.T) {
> ```

Looks like this test name is off. Did you mean to remove this version with `multiTestContext`?
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 189 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLDtWJ20CGfZq__g8XG:-MLDtWJ20CGfZq__g8XH:bot72ab) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d60b64af89c45cdebec183d2001003b28603051/pkg/kv/kvserver/replica_proposal_buf.go#L189)):*
> ```Go
> }
> 
> // setTestingKnobs should only be called immediately after Init(); it cannot be
> ```

nit: I would skip the unexported setter method and just assign directly like we do for the other testing knobs, but this is also fine.
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfTU7kAcuIXjgA8XLl:-MLDtie_-4l5ss_yHxPj:b11w1px) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf_test.go#L74)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

added test
</blockquote></details>

Did you want to add a test to this file? I think that would be the easiest way to unit test this change. It should be pretty satisfying to do.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @bdarnell, @knz, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/client_raft_test.go, line 1283 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLDt3lCAJD34XBhzyJp:-MLP1lgI356cXHuZeDst:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d60b64af89c45cdebec183d2001003b28603051/pkg/kv/kvserver/client_raft_test.go#L1283)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Looks like this test name is off. Did you mean to remove this version with `multiTestContext`?
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 189 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLDtWJ20CGfZq__g8XG:-MLPn-3VDUfNp0Tgd2Nw:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d60b64af89c45cdebec183d2001003b28603051/pkg/kv/kvserver/replica_proposal_buf.go#L189)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: I would skip the unexported setter method and just assign directly like we do for the other testing knobs, but this is also fine.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MIfTU7kAcuIXjgA8XLl:-MLPmvHB2_4SPnNu8cAU:bdc5xm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce45898fe977438f8706e9757ac5e947a4e5991/pkg/kv/kvserver/replica_proposal_buf_test.go#L74)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did you want to add a test to this file? I think that would be the easiest way to unit test this change. It should be pretty satisfying to do.
</blockquote></details>

done, PTAL


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> the new testing looks good.

> Reviewed 1 of 12 files at r8, 11 of 11 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-:-MLU0Krs-HcH7o8Uig8u:boxvbmf)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @knz, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 183 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLTzouU6q9ouHsPLA9Y:-MLTzouVBfW5vGXShGK_:bnpmfbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf.go#L183)):*
> ```Go
> }
> 
> // proposerRaft abstracts the propBuf's dependency on raft.RawNode, to help
> ```

nit: `*raft.RawNode`. 
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 41 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLU-ByFEM2fEYXrg9xw:-MLU-ByFEM2fEYXrg9xx:b-ejtadz) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf_test.go#L41)):*
> ```Go
> 	registered int
> 
> 	// If not nil, this can be a testRaftProposer used to mock the raft group
> ```

`s/testRaftProposer/testProposerRaft/`
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 54 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLU-Q7l-ne1xHFx7GoL:-MLU-Q7l-ne1xHFx7GoM:br0plix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf_test.go#L54)):*
> ```Go
> 
> func (t testProposerRaft) Step(raftpb.Message) error {
> 	return nil
> ```

Mind leaving me a TODO to capture this message and test against it? Same thing down in ProposeConfChange.
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 107 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLU-a6N94-KOcHlS4LU:-MLU-a6N94-KOcHlS4LV:b5xua3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf_test.go#L107)):*
> ```Go
> }
> 
> var _ proposerRaft = testProposerRaft{}
> ```

nit: either move this up near the rest of `testProposerRaft` or get rid of it. There's already static validation that this is true up above, so I'd lean towards the latter option.
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 383 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLU-ouW3hLUYMD0DVJ9:-MLU-ouXEOKG3AqN-evV:bds3i29) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf_test.go#L383)):*
> ```Go
> // followers. We want the leader to take the lease; see comments in
> // FlushLockedWithRaftGroup().
> func TestProposalBufferRejectLeaseAcqOnFollower(t *testing.T) {
> ```

Nice test!
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 428 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLU05NB19O0snF85lXg:-MLU05NB19O0snF85lXh:b-wlatk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf_test.go#L428)):*
> ```Go
> 			require.Equal(t, self, uint64(p.replicaID()))
> 
> 			var rejected []roachpb.ReplicaID
> ```

Why the slice? Do we expect multiple elements?
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 457 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLU099N9rPXyA8ffEYw:-MLU099N9rPXyA8ffEYx:bc2ac3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf_test.go#L457)):*
> ```Go
> 			if tc.expRejection {
> 				require.ElementsMatch(t, rejected, []roachpb.ReplicaID{roachpb.ReplicaID(tc.leader)})
> 			}
> ```

Do you want to assert that the slice is empty if `tc.expRejection`?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @bdarnell, @knz, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal_buf.go, line 183 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLTzouU6q9ouHsPLA9Y:-MLht4Te5eWq0Bk2FqI2:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf.go#L183)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `*raft.RawNode`. 
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 41 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLU-ByFEM2fEYXrg9xw:-MLhsejV6qlW52CTndYV:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf_test.go#L41)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/testRaftProposer/testProposerRaft/`
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 54 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLU-Q7l-ne1xHFx7GoL:-MLhspkP6ZIdlTXOHOY-:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf_test.go#L54)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Mind leaving me a TODO to capture this message and test against it? Same thing down in ProposeConfChange.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 107 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLU-a6N94-KOcHlS4LU:-MLhreHx-A0r9r9iML7K:b-7mmp82) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf_test.go#L107)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: either move this up near the rest of `testProposerRaft` or get rid of it. There's already static validation that this is true up above, so I'd lean towards the latter option.
</blockquote></details>

moved
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 428 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLU05NB19O0snF85lXg:-MLhsOny6zBJuGdi8JI-:b-n9tkhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf_test.go#L428)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why the slice? Do we expect multiple elements?
</blockquote></details>

changed
___
*[pkg/kv/kvserver/replica_proposal_buf_test.go, line 457 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-MLU099N9rPXyA8ffEYw:-MLhsPbA2pKqnBmUDea4:bbjfmbn) ([raw file](https://github.com/cockroachdb/cockroach/blob/af70483a1b069fcac27efef7188b0dfd9430274f/pkg/kv/kvserver/replica_proposal_buf_test.go#L457)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do you want to assert that the slice is empty if `tc.expRejection`?
</blockquote></details>

done - although it was already tested through the different `onRejectProposalWithRedirectLocked` we install in that case


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 11 of 11 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-:-MLp8Sz2582-44jrrbyS:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bdarnell, @knz, and @tbg)



<!-- Sent from Reviewable.io -->

andreimatei committed:
kvserver: improve a test

This test was performing a configuration change by talking directly to
one of the range's replicas. Instead, this change switches it to use an
Admin request which gets routed to the leaseholder. It's better if the
leaseholder coordinates the configuration change; a non-leaseholder can
error out because it might try to send snapshot before it applied the
preceding descriptor change locally (see #56596). The test seems to
currently work realiably, but, without this patch, becomes flaky with
the next commit which restricts who can take the lease for a range with
no lease.

Release note: None

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-:-MLxYobx5nP9rsLGK4dn:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bdarnell, @knz, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

andreimatei commented:
bors r-

knz commented:
i think you can let it go in now

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55148#-:-MM2Frj-9Xz6OI1XM58J:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bdarnell, @knz, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->
